File: backend/app/domain/slide_pages.py
"""编辑器内的整页操作：空白页内容与整页克隆。

只做纯数据变换（大纲页、blocks、布局树），落库与序号维护在
``services/deck_pages.py``。
"""

from __future__ import annotations

import uuid
from typing import Any

from app.domain.flex_edit import default_block_dict
from app.domain.flex_layout import FlexContainer, FlexLeaf, FlexNode
from app.domain.flex_normalize import normalize
from app.domain.outline import OutlinePage

# 空白页用最通用的要点版式：既能过大纲的布局校验，也是手写页最常见的结构
BLANK_LAYOUT_ID = "bullets"
BLANK_TITLE = "新页面"
_BLANK_OBJECTIVE = "补充这一页想讲清楚的事"
# OutlinePage 要求至少两条要点，占位文案同时也是给用户的填写提示
_BLANK_KEY_POINTS = ("待补充要点", "待补充要点")
# 与 title-bullets 预设一致的纵向权重
_TITLE_GROW = 0.45
_BODY_GROW = 1.5


def blank_outline_page() -> OutlinePage:
    """空白页对应的大纲页。

    slides 以 outline_page_id 为幂等键，且 ``sync_slides`` 会删掉大纲里不存在
    的页，所以编辑器插入的页必须在大纲里同时有一条记录。
    """
    return OutlinePage(
        title=BLANK_TITLE,
        objective=_BLANK_OBJECTIVE,
        key_points=list(_BLANK_KEY_POINTS),
        source_refs=[],
        layout_id=BLANK_LAYOUT_ID,
        page_role="content",
    )


def blank_slide_content() -> tuple[list[dict[str, Any]], FlexContainer]:
    """空白页的最小可编辑内容：一个标题加一组要点。

    这里直接写死一棵纵向树而不走预设挑选：预设按块类型组成打分，两块内容会被
    评到左右分栏之类的版式上，对一张待填写的空页毫无必要。
    """
    prefix = uuid.uuid4().hex[:12]
    title_id = f"{prefix}-title"
    body_id = f"{prefix}-body"
    blocks = [
        {**default_block_dict("text", title_id), "text": BLANK_TITLE},
        default_block_dict("bullets", body_id),
    ]
    tree = normalize(
        FlexContainer(
            type="column",
            id="root",
            children=[
                FlexLeaf(
                    id=f"leaf-{title_id}",
                    block_id=title_id,
                    grow=_TITLE_GROW,
                    text_style="title",
                ),
                FlexLeaf(
                    id=f"leaf-{body_id}",
                    block_id=body_id,
                    grow=_BODY_GROW,
                    text_style="bullet",
                ),
            ],
        )
    )
    return blocks, tree


def clone_slide_content(
    blocks: list[dict[str, Any]],
    layout_tree: dict[str, Any] | None,
) -> tuple[list[dict[str, Any]], dict[str, Any] | None]:
    """深拷贝整页内容，并把全部块 id 换成新的一批。

    两页各存一份 JSON，id 撞车本身写不坏数据，但编辑撤销栈与选中态都按
    block id 寻址，沿用旧 id 会让复制页与原页互相串。
    """
    renames = {str(block["id"]): uuid.uuid4().hex[:12] for block in blocks if block.get("id")}
    cloned = [_clone_block(block, renames) for block in blocks]
    if layout_tree is None:
        return cloned, None
    tree = _clone_node(FlexContainer.model_validate(layout_tree), renames)
    return cloned, tree.model_dump(mode="json")


def _clone_block(block: dict[str, Any], renames: dict[str, str]) -> dict[str, Any]:
    cloned = _deep_copy(block)
    old_id = str(block.get("id", ""))
    new_id = renames.get(old_id)
    if new_id is None:
        return cloned
    cloned["id"] = new_id
    # flex 约定 slot_id 与 id 同值；fixed 页的槽位名是布局定义的一部分，保持原样
    if str(block.get("slot_id", "")) == old_id:
        cloned["slot_id"] = new_id
    return cloned


def _clone_node(node: FlexNode, renames: dict[str, str]) -> FlexNode:
    if isinstance(node, FlexLeaf):
        new_id = renames.get(node.block_id)
        if new_id is None:
            return node.model_copy(deep=True)
        return node.model_copy(update={"block_id": new_id, "id": f"leaf-{new_id}"})
    children = [_clone_node(child, renames) for child in node.children]
    return node.model_copy(update={"children": children})


def _deep_copy(value: Any) -> Any:
    if isinstance(value, dict):
        return {key: _deep_copy(item) for key, item in value.items()}
    if isinstance(value, list):
        return [_deep_copy(item) for item in value]
    return value

File: java_backend/src/main/java/com/aippt/domain/content/SlidePages.java
package com.aippt.domain.content;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import com.aippt.domain.OutlinePage;
import com.aippt.domain.flex.FlexContainer;
import com.aippt.domain.flex.FlexLeaf;
import com.aippt.domain.flex.FlexNormalize;
import com.aippt.domain.flex.FlexTrees;
import com.aippt.shared.json.JsonMapperHolder;
import com.fasterxml.jackson.core.type.TypeReference;

public final class SlidePages {

    public static final String BLANK_LAYOUT_ID = "bullets";
    public static final String BLANK_TITLE = "新页面";
    private static final String BLANK_OBJECTIVE = "补充这一页想讲清楚的事";
    private static final List<String> BLANK_KEY_POINTS = List.of("待补充要点", "待补充要点");
    private static final double TITLE_GROW = 0.45;
    private static final double BODY_GROW = 1.5;

    public record BlankContent(List<Map<String, Object>> blocks, FlexContainer tree) {
    }

    public record ClonedContent(List<Map<String, Object>> blocks, Map<String, Object> layoutTree) {
    }

    private SlidePages() {
    }

    public static OutlinePage blankOutlinePage() {
        return new OutlinePage(
                UUID.randomUUID(),
                BLANK_TITLE,
                BLANK_OBJECTIVE,
                BLANK_KEY_POINTS,
                List.of(),
                BLANK_LAYOUT_ID,
                "content",
                null
        );
    }

    public static BlankContent blankSlideContent() {
        String prefix = UUID.randomUUID().toString().replace("-", "").substring(0, 12);
        String titleId = prefix + "-title";
        String bodyId = prefix + "-body";
        Map<String, Object> title = Blocks.defaultBlock("text", titleId);
        title.put("text", BLANK_TITLE);
        List<Map<String, Object>> blocks = List.of(title, Blocks.defaultBlock("bullets", bodyId));
        FlexContainer tree = FlexNormalize.normalize(FlexContainer.column(
                "root",
                List.of(
                        FlexLeaf.of("leaf-" + titleId, titleId, TITLE_GROW, "title"),
                        FlexLeaf.of("leaf-" + bodyId, bodyId, BODY_GROW, "bullet")
                ),
                16.0,
                1.0
        ));
        return new BlankContent(blocks, tree);
    }

    public static ClonedContent cloneSlideContent(List<Map<String, Object>> blocks, Map<String, Object> layoutTree) {
        Map<String, String> renames = new LinkedHashMap<>();
        if (blocks != null) {
            for (Map<String, Object> block : blocks) {
                String oldId = Blocks.id(block);
                if (!oldId.isBlank()) {
                    renames.put(oldId, UUID.randomUUID().toString().replace("-", "").substring(0, 12));
                }
            }
        }
        List<Map<String, Object>> cloned = new ArrayList<>();
        if (blocks != null) {
            for (Map<String, Object> block : blocks) {
                cloned.add(cloneBlock(block, renames));
            }
        }
        if (layoutTree == null) {
            return new ClonedContent(cloned, null);
        }
        FlexContainer tree = FlexTrees.cloneAndRename(FlexTrees.parse(layoutTree), renames);
        return new ClonedContent(cloned, tree.toMap());
    }

    public static Map<String, Object> outlinePageMap(OutlinePage page) {
        return JsonMapperHolder.MAPPER.convertValue(page, new TypeReference<>() {
        });
    }

    private static Map<String, Object> cloneBlock(Map<String, Object> block, Map<String, String> renames) {
        Map<String, Object> cloned = Blocks.deepCopy(block);
        String oldId = Blocks.id(block);
        String newId = renames.get(oldId);
        if (newId == null) {
            return cloned;
        }
        cloned.put("id", newId);
        if (oldId.equals(Blocks.slotId(block))) {
            cloned.put("slot_id", newId);
        }
        return cloned;
    }
}
